specification/sbi_atee.adoc

```diff
@@ -10,8 +39,1296 @@ shown in the figure above, the TEE Runtime (operating in VS-mode) supports
+in the range will result in a fatal page fault.
+
+[#table_sbi_tee_guest_flush_pages_errors]
+.TEE Guest Flush Pages
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code              | Description
+| SBI_SUCCESS             | The invalidation operation was successfully started.
+                          | This implies an exit to the host, and a subsequent resume of execution.
+| SBI_ERR_INVALID_ADDRESS | `tvm_gpa_addr` was invalid.
+| SBI_ERR_INVALID_PARAM   | `region_len` was invalid, or the range spanned a region 
+                            of another type.
+| SBI_ERR_FAILED          | The operation failed for unknown reasons.
+|===
+
+=== Function: TEE Guest Initiate Global Fence (FID #2)
+[source, C]
+-----
+struct sbiret sbi_tee_guest_global_fence(void);
+-----
+Initiates a TLB invalidation sequence for all pages marked for conversion via
+calls to `sbi_tee_host_convert_pages()`. The TLB invalidation sequence is completed
+when `sbi_guest_host_local_fence()` has been invoked on all other CPUs. An error is
+returned if a TLB invalidation sequence is already in progress.
+
+The possible error codes returned in `sbiret.error` are shown below.
+
+[#table_sbi_sbi_tee_guest_global_fence_errors]
+.TEE Host Initiate Fence
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code              | Description
+| SBI_SUCCESS             | The operation completed successfully.
+| SBI_ERR_ALREADY_STARTED | A fence operation is already in progress.
+| SBI_ERR_FAILED          | The operation failed for unknown reasons.
+|===
+
+=== Function: TEE Guest Local Fence (FID #3)
+[source, C]
+-----
+struct sbiret sbi_tee_guest_local_fence(void);
+-----
+Invalidates TLB entries for all pages pending conversion by an in-progress TLB
+invalidation operation on the local CPU.
+
+The possible error codes returned in `sbiret.error` are shown below.
+
+[#table_sbi_tee_guest_local_fence_errors]
+.TEE Guest Local Fence
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code            | Description
+| SBI_SUCCESS           | The operation completed successfully.
+| SBI_ERR_FAILED        | The operation failed for unknown reasons.
+|===
+
+=== Function: TEE Guest Share Memory Region (FID #4)
+[source, C]
+-------
+struct sbiret sbi_tee_guest_share_memory_region(unsigned long tvm_gpa_addr,
+                                                unsigned long region_len);
+-------
+Marks the range of TVM physical address space starting at `tvm_gpa_addr` as reserved for the mapping of non-confidential memory. The entire range must reside in a region of
+type `CONFIDENTIAL_REGION`, and all pages must have been previously unmapped and flushed (using `sbi_tee_guest_flush_pages()`, followed by a fence).
+
+Both `tvm_gpa_addr` and `region_len` must be 4kB-aligned. This call will result in an exit to the
+host on success. Following successful completion, the TVM can access all any page in the range.
```

Comment:
This needs to be re-written to reflect what was actually implemented. I'd suggest something like this:

"""
Converts the requested range of TVM guest physical address space from confidential to shared. The requested range must lie within an existing region of confidential address space, and may or may not be populated.

If the region of address space is populated, the TSM invalidates the pages and marks the region as pending conversion to shared. The host must complete a TVM TLB invalidation sequence, initiated by `tee_host_tvm_initiate_fence()`, in order to complete the conversion. The calling TVM vCPU is considered blocked until the conversion is completed; attempts to run it with `tee_host_tvm_run()` will fail with <some error code>. Any guest page faults taken by other TVM vCPUs in this region prior to completion of the conversion are considered fatal. The host may not insert any pages in the region prior to completion of the conversion. Upon completion, the host may reclaim the confidential pages that were previously mapped in the region using `tee_host_tsm_reclaim_pages()` and may insert shared pages into the region using `tee_host_tvm_add_shared_pages()`.

If the range of address space is completely unpopulated, the region is immediately converted to shared and the host may insert shared pages.
"""

cc: @rsahita

Reply:
this LGTM - I can merge this change into the writeup.